File: app/services/dadata.py
"""DaData integration for company lookup."""

from __future__ import annotations

from dataclasses import dataclass
import asyncio
from typing import Any

import httpx

from app.utils.logging import get_logger


DADATA_FIND_BY_ID_URL = "https://suggestions.dadata.ru/suggestions/api/4_1/rs/findById/party"
MAX_RETRIES = 2

logger = get_logger(__name__)


@dataclass(slots=True)
class CompanyProfile:
    """Structured company profile from DaData."""

    name: str | None
    inn: str | None
    ogrn: str | None
    kpp: str | None
    address: str | None
    director: str | None
    status: str | None
    registration_date: str | None
    mass_address: bool | None
    mass_director: bool | None


class DadataError(RuntimeError):
    """Raised for DaData API errors."""


async def fetch_company_profile(
    inn: str, token: str, secret: str | None, timeout_seconds: float
) -> CompanyProfile:
    """Fetch company profile from DaData by INN."""

    headers = {
        "Authorization": f"Token {token}",
        "Accept": "application/json",
        "Content-Type": "application/json",
    }
    if secret:
        headers["X-Secret"] = secret

    payload = {"query": inn}
    timeout = httpx.Timeout(timeout_seconds)
    last_error: Exception | None = None

    for attempt in range(MAX_RETRIES + 1):
        try:
            async with httpx.AsyncClient(timeout=timeout) as client:
                response = await client.post(
                    DADATA_FIND_BY_ID_URL, headers=headers, json=payload
                )
            response.raise_for_status()
            data = response.json()
            suggestions = data.get("suggestions") or []
            if not suggestions:
                raise DadataError("DaData returned empty suggestions list")

            suggestion = suggestions[0]
            data_fields: dict[str, Any] = suggestion.get("data") or {}
            # ASSUMPTION: DaData uses 'name.full_with_opf' and 'address.unrestricted_value'.
            name = (data_fields.get("name") or {}).get("full_with_opf")
            address = (data_fields.get("address") or {}).get("unrestricted_value")
            management = data_fields.get("management") or {}
            director = management.get("name")

            status = (data_fields.get("state") or {}).get("status")
            registration_date = (data_fields.get("state") or {}).get("registration_date")

            return CompanyProfile(
                name=name,
                inn=data_fields.get("inn"),
                ogrn=data_fields.get("ogrn"),
                kpp=data_fields.get("kpp"),
                address=address,
                director=director,
                status=status,
                registration_date=registration_date,
                mass_address=data_fields.get("mass_address"),
                mass_director=data_fields.get("mass_director"),
            )
        except (httpx.RequestError, httpx.HTTPStatusError, DadataError) as exc:
            last_error = exc
            logger.warning(
                "DaData request failed",
                extra={
                    "module": "dadata",
                    "status": "retry" if attempt < MAX_RETRIES else "failed",
                    "error_type": type(exc).__name__,
                },
            )
            if attempt < MAX_RETRIES:
                await asyncio.sleep(0.5 * (attempt + 1))
                continue
            break

    raise DadataError("Failed to fetch DaData profile") from last_error

File: src/services/dadata.js
const axios = require('axios');
const { config, requireEnv } = require('../config');
const { logError } = require('../utils/logger');

const DADATA_TIMEOUT_MS = Math.round(config.httpTimeoutSeconds * 1000);

/**
 * @param {string} inn
 * @param {{ count?: number, timeoutMs?: number }} [options]
 * @returns {Promise<unknown>}
 */
async function findPartyByInn(inn, options = {}) {
  const apiKey = requireEnv('DADATA_API_KEY');
  const secret = requireEnv('DADATA_SECRET_KEY');
  const count = options.count ?? config.dadataCount;
  const timeoutMs = options.timeoutMs ?? DADATA_TIMEOUT_MS;

  const response = await axios.post(
    config.dadataUrl,
    { query: String(inn), count },
    {
      timeout: timeoutMs,
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
        Authorization: `Token ${apiKey}`,
        'X-Secret': secret,
      },
    },
  );

  return response.data;
}

/**
 * @param {unknown} dadataResponse
 * @returns {null | {
 *  name: string | null,
 *  inn: string | null,
 *  kpp: string | null,
 *  ogrn: string | null,
 *  status: string | null,
 *  okved: string | null,
 *  address: string | null,
 *  management: unknown,
 *  phones: unknown[],
 *  emails: unknown[]
 * }}
 */
function normalizeParty(dadataResponse) {
  const suggestion = dadataResponse?.suggestions?.[0];
  if (!suggestion) return null;

  const data = suggestion.data || {};
  return {
    name: data?.name?.short_with_opf || suggestion.value || null,
    inn: data.inn || null,
    kpp: data.kpp || null,
    ogrn: data.ogrn || null,
    status: data?.state?.status || null,
    okved: data.okved || null,
    address: data?.address?.unrestricted_value || data?.address?.value || null,
    management: data.management || null,
    phones: Array.isArray(data.phones) ? data.phones : [],
    emails: Array.isArray(data.emails) ? data.emails : [],
  };
}

/**
 * @param {Error} error
 * @param {{ operation: string, inn: string }} context
 */
function logDadataError(error, context) {
  const status = error?.response?.status;
  logError('DaData request failed', {
    operation: context.operation,
    result: 'error',
    inn: context.inn,
    status,
  });
}

module.exports = {
  findPartyByInn,
  normalizeParty,
  logDadataError,
};
